File: batou-rules/testdata/fixtures/go/vulnerable/zip_slip.go
package handler

import (
	"archive/zip"
	"io"
	"os"
	"path/filepath"
)

func extractZip(zipPath, destDir string) error {
	r, err := zip.OpenReader(zipPath)
	if err != nil {
		return err
	}
	defer r.Close()

	for _, f := range r.File {
		// VULNERABLE: using zip entry name directly in filepath.Join without validation
		destPath := filepath.Join(destDir, f.Name)
		os.MkdirAll(filepath.Dir(destPath), 0755)

		rc, err := f.Open()
		if err != nil {
			return err
		}

		outFile, err := os.Create(destPath)
		if err != nil {
			rc.Close()
			return err
		}

		io.Copy(outFile, rc)
		outFile.Close()
		rc.Close()
	}
	return nil
}

File: batou-rules/testdata/fixtures/python/vulnerable/zip_slip.py
import tarfile
import os

def extract_tar(tar_path, dest_dir):
    """Vulnerable: extractall without members filter or safe filter."""
    tf = tarfile.open(tar_path)
    tf.extractall(dest_dir)
    tf.close()

def extract_tar_manual(tar_path, dest_dir):
    """Vulnerable: manual extraction with path join using entry name."""
    tf = tarfile.open(tar_path)
    for member in tf.getmembers():
        dest_path = os.path.join(dest_dir, member.name)
        tf.extract(member, dest_dir)
    tf.close()
